util/vlanips.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2024 Akamai Technologies, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"net/netip"
+	"slices"
+	"sync"
+)
+
+var (
+	vlanIPsMu sync.Mutex
+	// vlanIPsMap stores clusterName and a list of VlanIPs assigned to that cluster
+	vlanIPsMap  = make(map[string][]string, 0)
+	vlanIPRange = "10.0.0.0/8"
+)
+
+// GetNextVlanIP returns the next available IP for a cluster
+func GetNextVlanIP(clusterName, namespace string) string {
+	vlanIPsMu.Lock()
+	defer vlanIPsMu.Unlock()
```

Comment:
Instead of having one global lock, can we use two locks here? One for global map and second for individual cluster? I see the current implementation becoming a bottleneck if the cluster size increases as then we would be iterating for a longer duration to find the next free ip. I have a similar example here: https://github.com/linode/cluster-api-provider-linode/blob/main/util/ratelimits.go#L32-L43

Reply:
once a IP is assigned, its going to end up having to update the global map - so its going to have to lock the global map regardless, if i'm not missing something

Reply:
pushed up a change with a per-cluster lock - let me know what you think